### DISCUSSION TITLE
API Documentation Issue: Creating Attachments

### DISCUSSION THREAD

**Discussion by NoahMarch** [outline/outline#7712] [Self Hosting]
Hello all,

The documentation for the API regarding file uploads via the API doesn't seem to be complete.
https://www.getoutline.com/developers#tag/attachments/POST/attachments.create
After using this and encountering a 400 error {"ok":false,"error":"validation_error","status":400,"message":"preset: Required"}, we figured out that adding "preset":"documentAttachment" to the JSON body resolves the issue.

However, the next step — how to actually upload the file (api/files.create) — is missing entirely from the API documentation.
We used:

"key": "TheKeyYouGetFromAttachmentsCreate"
and as binary data, "file" with the file data.
This successfully created the file. However, during /attachments.create, we specified the documentId with the ID of a document we had created previously. The file is uploaded and viewable via the redirect URL you receive from the /attachments.create response, but the file itself isn't attached (it doesn't show up on the page) to the document, as expected when setting the "Identifier for the associated document, if any."

In the end, the question remains: How can we upload and attach the file to a document via the API?

_version 0.80.2 hosted via Cloudron_

**Comment by tommoor** [comment]
Hi @NoahMarch there is no way to attach a file to the document via API directly but it can be achieve in three steps, this is what the client does:

1. Generate the attachment record and upload params (`attachments.create`)
2. Upload the file for the attachment using the parameters from 1.
3. Use `documents.update` to mutate the content of the document and create a link to the file

**Reply by NoahMarch** [reply]
Thanks alot. However im not sure if this i correct:


If i retrive a document it seems like Attachments are `[Name](/api/attachmants.redirect?id=ABC123)`. However if i send a http request with the Following json to `documents.update`:
```
{ 
"id": "IdOfDocument",
"text": "[Literatur.pdf](/api/attachments.redirect?id=IdOfAttachement)",
"publish": true,
"append": false
}
```
In the response i find the follwoing entry with an added space for text:
`[ Literatur.pdf](/api/attachments.redirect?id=IdOfAttachement)`

The attachment is displayed as the following:
![image](https://github.com/user-attachments/assets/eb525356-4cae-4a82-9c58-ffb9e59f9e93)
Its clickable and opens the correct attachment, however the name is not displayed correct - maybe that has something to do with the added Space? 


Also if using this workflow, what ist the beneift of defining documentId in the `attachments.create` process?

**Reply by tommoor** [reply]
There is no Markdown standard for such things, the following format is used for attachments:

`[title sizeInBytes](url)`

If you add the size then it should display correctly

**Reply by tommoor** [reply]
> Also if using this workflow, what ist the beneift of defining documentId in the attachments.create process?

None currently, this is vestigial – however we might have a way to browse attachments in the future and this would offer a signal as to where it was originally uploaded.
